acp/acp.go

```diff
@@ -0,0 +1,113 @@
+// Note: sourcehub acp module will likely need the remote implementation.
+type ACPModule interface {
+
+	// Start initializes the acp module, with persistence data stored under the given folder.
+	//
+	// If previously started in the same path, then recovers / reloads (unless used in-memory).
+	Start(context.Context, string) error
+
+	// Close closes the resources in use by the acp module.
+	Close() error
+
+	// AddPolicy attempts to add the given policy. Detects the format of the policy automatically
+	// by assuming YAML format if JSON validation fails. Upon success a policyID is returned,
+	// otherwise returns error.
+	//
+	// A policy can not be added without a valid creator.
+	AddPolicy(ctx context.Context, creator string, policy string) (string, error)
+
+	// ValidateResourceExistsOnValidDPI consumes a policyID and resource to validate that the
+	// the following conditions are met:
+	// - policyID is a valid argument.
+	// - resource name is a valid argument.
+	// - policyID corresponds to a valid policy that is registered on acp module.
+	// - resource name exists on that registered policy.
+	// - resource is DPI compliant (more information on DPI below).
+	//
+	// An error is returned upon validation failure (i.e. if above rules not met).
+	//
+	// To be a DPI compliant resource, these rules MUST be met:
+	// - The resource must exist.
+	// - The resource must atleast have the required registerer relation under `relations`.
+	// - The resource must atleast have all the required permissions under `permissions`.
+	// - Every required permission must have the required registerer relation in `expr`.
+	// - The required registerer relation must be the leading (first) relation in `expr`.
+	//
+	// Learn more about the DefraDB Policy Interface [DPI](/acp/DPI.md)
+	// Learn more about DefraDB ACP Terminologies [TERMINOLOGY](/acp/TERMINOLOGY.md)
+	ValidateResourceExistsOnValidDPI(ctx context.Context, policyID, resource string) error
+
+	// RegisterDocCreation registers the document (object) to have access control.
+	// No error is returned upon successful registering of a document.
+	//
+	// Note(s):
+	// - This should be used upon document creation only.
+	// - Some documents might be created without an identity signature so they would have public access.
+	// - creator here is the actorID, which will be the signature identity if it exists.
+	// - resource here is the resource object name (likely collection name).
+	// - docID here is the object identifier.
+	RegisterDocCreation(ctx context.Context, creator, policyID, resource, docID string) error
+
+	// IsDocRegistered returns true if the check was successfull and the document was found to be registered
+	// with ACP, if the check was successful and the document was not found to be registered then return false.
+	// Otherwise if check failed then an error is returned (and the boolean result should not be used).
```

Comment:
suggestion: This feels a very long-winded way of writing this. Maybe the below would be a little better?

```
IsDocRegistered returns true the document was found to be registered, otherwise false. If check failed then an error and false will be returned.
```

Same goes for `CheckDocAccess`

Reply:
I will think and revisit this suggestion, if I like it will do it.

- [ ] Revisit and see if I want to accept the suggestion

Reply:
- [ ] Fix `IsDocRegistered` documentation once the sourcehub fix is merged in.